### ISSUE TITLE
ndimage: segfault on incorrect parameters to correlate (Trac #295)

### ISSUE THREAD

**Issue by scipy-gitbot** [scipy/scipy#822]
_Original ticket http://projects.scipy.org/scipy/ticket/295 on 2006-10-25 by @stefanv, assigned to @stefanv._

```
In [2]: import scipy.ndimage as ndi

In [3]: ndi.correlate([0,1,2,3,4,5],[0,1,2],origin=[2])
```

**Comment by laemp** [comment]
I fail to see the problem.
`ndimage.correlate([0,1,2,3,4,5],[1,1,2,0],origin=2)`
should not be allowed because only values `-2,-1,0,1` for `origin` are valid in this case and they don't cause trouble.
So changing the `>` to `>=` and moving the enclosing for-loop before the
`if convolution:`
in line no. 531 also solves the problem for the `ndimage.convolve` function.
Would this not be the desired behaviour?

**Comment by stilley2** [comment]
Just to note I recently encountered this bug using the convolve function with version 0.19.0. 

    from scipy.ndimage import convolve
    k = [1.0, 1.0, 1.0]
    x = list(range(10))
    convolve(x, k, origin=-1)   # works
    convolve(x, k, origin=-3)   # ValueError: invalid origin
    convolve(x, k, origin=-2)   # segfault

**Comment by WarrenWeckesser** [comment]
@stilley2 Thanks for the report.  I also get the seg. fault (on Mac OS X 10.9.5) using the current master branch of scipy:

```
In [1]: import numpy as np

In [2]: np.__version__
Out[2]: '1.13.1'

In [3]: import scipy

In [4]: scipy.__version__
Out[4]: '1.1.0.dev0+8a994ca'

In [5]: from scipy.ndimage import convolve
   ...: k = [1.0, 1.0, 1.0]
   ...: x = list(range(10))
   ...: 

In [6]: convolve(x, k, origin=-2)
Segmentation fault: 11
```

**Comment by WarrenWeckesser** [comment]
I created a pull request that fixes this issue.  As already mentioned, the fix is to simply change an inequality in the validation of the `origin` argument.  In the pull request, I reformulated that expression, but the net effect is the same as the change mentioned above.

Josef said this example should be accepted as valid

    ndimage.correlate1d([0,1,2,3,4,5],[1,1,2,0],origin=2)

but in fact, that is not a valid origin.  If we think of the "sliding window" in which `weights` starts at the left and slides over `input` to the right, the code requires that `origin` be constrained so that starting position of `weights` has at least one element of overlap with `input` (the original input array, not the extended array).  When `weights` has length 4, the default starting position of `weights` has `weights[2]` lined up with `input[0]`.  A positive value of `origin` moves the start position of `weights` to the left, and `origin=2` results in no overlap between `weights` and `input`.  I suspect it was just luck that the code worked for Josef.  When I run that example with the master branch, the Python interpreter crashes.
